### PR TITLE
Self-critical sequence training

### PR DESCRIPTION
**Pull request by jlibovicky**
This PR introduces self-critical reinforcement learning for recurrent decoding with sentence-level BLEU score and GLEU score.

Copy of #426.

### CODE REVIEW

**Comment by loveJasmine** [comment]
https://github.com/ufal/neuralmonkey/issues/479

**Comment by loveJasmine** [comment]
@jlibovicky  and other
would you give me some advices for the ini config for the chatbot/NMT task?

here is my ini:

[train_data]
class=dataset.load_dataset_from_files
s_source="../andy_data970201/tr.q"   ; about 9000000 line
s_target="../andy_data970201/tr.a"

[val_data]
class=dataset.load_dataset_from_files
s_source="../andy_data970201/va.q"   ; about 50000 line
s_target="../andy_data970201/va.a"

[shared_vocabulary]
class=vocabulary.from_dataset
max_size=8000

[source_vocabulary]
class=vocabulary.from_dataset
datasets=[<train_data>]
series_ids=["source"]
max_size=8000

[target_vocabulary]
class=vocabulary.from_dataset
datasets=[<train_data>]
series_ids=["target"]
max_size=8000

[encoder]
class=encoders.sentence_encoder.SentenceEncoder
name="sentence_encoder"
rnn_size=300
max_input_len=50
embedding_size=300
dropout_keep_prob=0.8
attention_type=decoding_function.Attention
data_id="source"
vocabulary=<source_vocabulary>

[decoder]
class=decoders.decoder.Decoder
name="decoder"
encoders=[<encoder>]
rnn_size=256
embedding_size=300
dropout_keep_prob=0.8
use_attention=True
data_id="target"
vocabulary=<target_vocabulary>
max_output_len=50

[self_critical]
class=trainers.self_critical_objective.self_critical_objective
decoder=<decoder>
reward_function=trainers.self_critical_objective.sentence_bleu
weight=0.5

[cross_entropy]
class=trainers.cross_entropy_trainer.xent_objective
decoder=<decoder>
weight=0.5

[trainer]
class=trainers.generic_trainer.GenericTrainer
objectives=[<cross_entropy>,<self_critical>]
l2_weight=1.0e-8
clip_norm=1.0

[runner]
class=runners.runner.GreedyRunner
decoder=<decoder>
output_series="series_named_greedy"

[bleu]
class=evaluators.bleu.BLEUEvaluator
name="BLEU-4"

[tf_manager]
class=tf_manager.TensorFlowManager
num_threads=4
num_sessions=1
minimize_metric=False
save_n_best=3

[main]
name="deepone with self-critical training"
output="deepone_out-self-critical"
runners=[<runner>]
tf_manager=<tf_manager>
trainer=<trainer>
train_dataset=<train_data>
val_dataset=<val_data>
evaluation=[("series_named_greedy", "target", <bleu>), ("series_named_greedy", "target", evaluators.ter.TER)]
batch_size=128
runners_batch_size=128
epochs=50
validation_period="120m"
random_seed=1234
logging_period="10m"

It takes 5G memory in my GPU, but it takes almost 3 day still undone one epoch,
I try the 256 for both batch_size &  runners_batch_size,  it report Out of Memory on M40(12G) GPU

**Comment by jlibovicky** [comment]
Self-critical training is very slow because it computes the reward function on CPU. If you want to use it, you should first pre-train your model using cross-entropy and fine-tune it.

**Comment by loveJasmine** [comment]
@jlibovicky 
as you told "computes the reward function on CPU", if I got a power multi core CPU, maybe I could utilize multithread to speeding up 
is this the right setting:

[tf_manager]
class=tf_manager.TensorFlowManager
num_threads=50
num_sessions=1
minimize_metric=False
save_n_best=3

however, I don't see the CPU occupation exceed 200%, 
but actually I can make it to more than 5000% in some program
